src/query/vacuum.c

```diff
@@ -2974,10 +2977,30 @@ vacuum_master_task::execute (cubthread::entry &thread_ref)
       return;
     }
 
+  if (!ts_Gl->is_page_server_connected()) 
+  {
+    // have to be connected to get mvcc status of PTSes.
+    return;
+  }
+
   PERF_UTIME_TRACKER_START (&thread_ref, &perf_tracker);
 
   m_oldest_visible_mvccid = log_Gl.mvcc_table.update_global_oldest_visible ();
   vacuum_er_log (VACUUM_ER_LOG_MASTER, "update oldest_visible = %lld", (long long int) m_oldest_visible_mvccid);
+ 
+  /* TODO temporary logging. The global one will be computed taking both into account, and the vacuum runs */ 
+  MVCCID global_pts_oldest_visible_mvccid = get_active_tran_server_ptr ()->get_oldest_active_mvccid_from_page_server ();
```

Comment:
I also don't know how this should be handled. Because there are three scenarios where vacuum is active:
- monolithic server
- scalability ATS with Local Storage (a configuration mostly used for debugging & validation, and which should still be supported, but which is used *without* PTS's)
- scalability ATS proper with Remote Storage

Given the above, I think we should only request this info from PS when `is_tran_server_with_remote_storage`.
And some [manual] tests should be performed on the monolithic server, preferably before merge if possible.

Reply:
Thank you. I didn't consider every scenario. I'll keep those scenarios in mind. After some manual tests, I'll re-request your review. By the way, to be clear, what you mean by `monolithic server` is ATS with Local Storage without PS?

From what I understand, the configurations are:
1. ATS + PS [+ PTS(es)] (normal scalability)
2. (1) + local stroage
3. ATS + PS on the same node (no PTS)
4. ATS + local storage (no PTS, no PS) (monolithic?)

Anything I'm missing? You can just link the related Jira issues.

Reply:
I decided to call `monolithic server` what is now in `develop` (ie: without scalability/LETS).
I wanted to call it `stand-alone` but in CUBRID standalone means the `-SA` option. :)

Reply:
>     1. ATS + PS [+ PTS(es)] (normal scalability)
>     2. (1) + local stroage
>     3. ATS + PS on the same node (no PTS)
>     4. ATS + local storage (no PTS, no PS) (monolithic?)

1) and 3) are the same.
In essence `ATS (remote storage) + ....` because it does not matter where the PS executes and stores its data, ATS will still retrieve heap & log pages from another process (PS).

Reply:
Thank you for the clarification. One more thing: `(1) + local storage` is not the case? I thought that the local storage can be attached independently with the configuration. The comment from `pgbuf_read_page_from_file_or_page_server ` also says like that, I guess (local, connection to PS):
```
// Server type --- Storage Config Type --- Data Type --- Connection to PS --- Local Read --- Remote Read
//     PS                Local            Perm & Temp          Never              Yes            No
// -----------------------------------------------------------------------------------------------------
//     TS                Local                Perm              Yes               Yes           Yes
//                                                               No               Yes            No
//                                            Temp             Yes/No             Yes            No
//                       Remote               Perm              Yes                No           Yes
//                                            Temp              Yes               Yes            No
```
But, it fails to start a ATS with local storage when there is a PS. 

If the ` `(1) + local storage` is possible, I think we should use `is_page_server_connected()`.

Reply:
And, I guess we should use `is_page_server_connected()` in any case because it works only when the PS is connected. Even when `is_tran_server_with_remote_storage()` is true, the page server possibly hasn't been connected yet depending on the initialization order.

Reply:
> uess we should use `is_page_server_connected()` in any case because it works only when the PS is connected. Even when `is_tran_server_with_remote_storage()` is true, the page server possibly hasn't been connected yet depending on the initialization

You have a point with this, but I still think `is_tran_server_with_remote_storage` is the more conservative approach:
- indeed, `is_page_server_connected` implies `is_tran_server_with_remote_storage`
- but `is_page_server_connected` is more of a technical prerequisite in the sense that:
  - if `is_tran_server_with_remote_storage` is true and `is_page_server_connected` is false, nothing should work and the TS should just log the error and crash/shutdown/refuse to continue

In a broader sense, `is_page_server_connected` should not even be used outside of `tran_server` (it is more of an implementation detail); visible from the current state where:
  - `is_page_server_connected` is only used 3 times outside `tran_server`
  - `is_tran_server_with_remote_storage` is used ~80 times all over the place

Reply:
> In a broader sense, `is_page_server_connected` should not even be used outside of `tran_server` (it is more of an implementation detail); visible from the current state where:

Or, better put, `is_page_server_connected` should be a last resort check without which operation cannot continue - like it is now where it is used to check right before an interaction between TS and PS (page retrieval).

In the future, I assume there will be a keep-alive type of communicaction either between servers or between a server and master. And, with that, the state of the `is_page_server_connected` will be updated by the keep-alive state and its check will not be as superfluous as it is now.

Reply:
Always thank you for your elaboration.

Reply:
To summarize. The following scenarios need to be covered:
- monolithic server (what is now in develop, without LETS)
  - vacuum needs to work as it was until now because of tests in various test suites that must continue to pass
- scalability ATS with Local Storage (a configuration mostly used for debugging & validation, and which should still be supported, but which is used without PTS's)
  - this is mostly a debug configuration where ATS is connected to a remote PS but also has its own local storage
  - the configuration is used to validate the transactional log replication mechanism
  - there are some tests implemented for this configuration but **no PTS's**
  - in this cas vacuum must work as for the monolithic scenario - ie: no requests for oldest active mvccid to the PS
- scalability ATS proper with Remote Storage
  - in this case vacuum must take the value(s) reported by PS into calculation

Given the above, I think `is_tran_server_with_remote_storage` fits the bill.

@wrlawodms Please add some variation of this comment as the rationale for the changes in `vacuum_master_task::execute` function. Thanks